File: 322-coin-change/322-coin-change.go
func coinChange(coins []int, amount int) int {
  if amount == 0 { return 0 }
  ans := make([]int, amount+1)
  min := 13
  for _, v := range coins {
    if v <= amount {
      ans[v] = 1
      if v == amount {
        return 1
      }
    }
    if v < min {
      min = v
    }
  }

  for i := min+1; i <= amount; i++ {
    if ans[i] > 0 { continue }
    min = amount+1
    for _, v := range coins {
      if v < i && ans[i-v] > 0 {
        if ans[i-v] < min {
          min = ans[i-v]
        }
      }
    }
    if min != amount+1 {
      ans[i] = min + 1
    }
  }
  if ans[amount] == 0 {
    return -1
  }
  return ans[amount]
}

File: 322-coin-change/322-coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
  const dp = Array.apply(null, Array(amount+1)).map(_ => Infinity)
  dp[0] = 0;
  for (i = 1; i <= amount; i++) {
    for (let coin of coins) {
      if(i-coin >= 0) { 
              
        dp[i] = Math.min(dp[i], dp[i-coin]+1);
      }
    }
       
  }
  if (dp[amount] == Infinity) {
    return -1
  } 
  return dp[amount];
};